### PR TITLE
Different type of task node

### PR DESCRIPTION
**Pull request by KKRainbow**

### CODE REVIEW

File: examples/charactercount/example.go [hunk 1]
```diff
@@ -29,6 +31,14 @@ type reverseReduce struct {
 	mapred.ReducerCommon
 }
 
+type outputText struct {
+	mapred.FilterCommon
+}
+
+func (*outputText) Filter(k, v interface{}, output io.Writer) {
+	output.Write([]byte(fmt.Sprintln(k, v)))
```
Comment (Lcch) [inline examples/charactercount/example.go]: Should we handle write exception? Should return error,

File: executor/filter_executor.go [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+}
+
+func NewFilterExecutor() *FilterExecutor {
+	return &FilterExecutor{}
+}
+
+func (e *FilterExecutor) isTargetFor(n jobgraph.TaskNode) bool {
+	_, ok := n.(*jobgraph.FilterNode)
+	return ok
+}
```
Comment (Lcch) [inline executor/filter_executor.go]: formate the code. Add a blank line after the funciton

File: executor/filter_executor.go [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+	// map
+	waitc := make(chan struct{})
+	inputKV := make(chan *kmrpb.KV, 1024)
+	filter.Init()
+	keyClass, valueClass := filter.GetInputKeyTypeConverter(), filter.GetInputValueTypeConverter()
+	go func() {
+		for kvpair := range inputKV {
+			key := keyClass.FromBytes(kvpair.Key)
+			value := valueClass.FromBytes(kvpair.Value)
+			filter.Filter(key, value, writer)
```
Comment (Lcch) [inline executor/filter_executor.go]: handle error exception?

File: executor/filter_executor.go [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+		}
+		close(waitc)
+	}()
+	for rr.HasNext() {
+		inputKV <- RecordToKV(rr.Pop())
```
Comment (Lcch) [inline executor/filter_executor.go]: Do not need to create another go routine for processing filter. Should put the handling function inside for loop.

File: master/scheduler.go [hunk 5]
```diff
@@ -179,33 +153,33 @@ func (s *Scheduler) StartSchedule(visitor EventHandler) error {
...
-						s.phaseMap[processingJob] = mapPhase
-						tasks = &processingJob.mapTasks
-					} else {
+					taskNode := processingJob.taskNode
+					jobPhase := s.phaseMap[processingJob]
+					if s.phaseFinishedCnt[processingJob][jobPhase] == taskNode.GetTaskCountOfPhase(jobPhase) {
+						jobPhase++
+						s.phaseMap[processingJob] = jobPhase
+					}
+					if jobPhase >= taskNode.GetPhaseCount() {
```
Comment (Lcch) [inline master/scheduler.go]: After if s.phaseFinishedCnt[processingJob][jobPhase] == taskNode.GetTaskCountOfPhase(jobPhase). jobPhase will increase, therefore, jobPhase will >= taskNode.GetPhaseCount() ?

File: mapred/other.go [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+// GetInputValueTypeConverter get input value type converter
+func (tb *InputTypeKeyConverters) GetInputValueTypeConverter() TypeConverter {
+	return tb.InputValueTypeConverter
+}
+
+type Filter interface {
+	InputTypeConverter
+	Filter(k, v interface{}, writer io.Writer)
+	Init()
+	After()
```
Comment (Lcch) [inline mapred/other.go]: After() is never be used on FilterExector

File: master/scheduler.go (hunk 5 above)
Comment (KKRainbow) [reply master/scheduler.go]: In ReportJob, when jobPhase == taskNode.GetPhaseCount(), this job will be marked finished and removed

File: mapred/other.go (hunk 6 above)
Comment (KKRainbow) [reply mapred/other.go]: Forgot it = =

File: master/scheduler.go (hunk 5 above)
Comment (Lcch) [reply master/scheduler.go]: get it

File: executor/filter_executor.go (hunk 4 above)
Comment (Lcch) [reply executor/filter_executor.go]: forget it

**Pull request merged**